lib/api/apiUtils/object/abortMultipartUpload.js

```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
                     }
                     cb();
                 });
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
+        },
+        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD, next) {
+            if (!objectMD || metadataValMPUparams.uploadId !== objectMD.uploadId) {
+                return next(null, mpuBucket, storedParts, destBucket, objectMD);
+            }
+            // In case there has been an error during cleanup after a complete MPU
+            // (e.g. failure to delete MPU MD in shadow bucket),
+            // we need to ensure that the MPU metadata is deleted.
+            log.info('Object has existing metadata, deleting them', {
+                method: 'abortMultipartUpload',
+                bucketName,
+                objectKey,
+            });
+            return metadata.deleteObjectMD(bucketName, objectKey, undefined, log, err => {
+                if (err) {
+                    log.error('error deleting object metadata', { error: err });
+                }
+                return next(err, mpuBucket, storedParts, destBucket, objectMD);
+            });
+        },
+        function deleteObjectData(mpuBucket, storedParts, destBucket, objectMD, next) {
```

Comment:
similar to the "parts", shouldn't we first delete data, then metadata? otherwise we can get in a bad state, where we don't know of the parts anymore...

this has the added benefit of allowing to remove all parts in a single stage, just combine the MPU parts with the object parts (if any)...

(note: if we decide to first delete the ObjectMD, to avoid having an object pointing to delete data, we need to do it before removing the MPU parts data, as they may be refereced ; and this allows deleting in a single step...)

Reply:
If I understand correctly, I think we prefer having "data orphans" instead of "metadata orphans" because in case of a "metadata orphans" user can think of a data loss because they can list and head the object but not get it..

This has been discussed here with jonathan
https://github.com/scality/cloudserver/pull/5691#discussion_r1841460336

Reply:
if we prefer having data orphans, then we need to delete the (object) metadata first : i.e. delete object metadata before the part's data (i.e. before `deleteData()` function), since the parts are likely used in the object as well...

...and doing that we don't need to add a separate step to remove the remaining object's data: we simply need to remove both the parts' data and object data in the `deleteDate()` phase.